### PR TITLE
Add pytorch version for DeCoAr code and model

### PR DESCRIPTION
**Pull request by lingss0918**
- Replace Mxnet code with Pytroch code for Decoar
- Add upstream model code for Decoar 2
- The model download path included

### CODE REVIEW

**Comment by leo19941227** [comment]
Hi!

Thanks for the PR. The `decoar` upstream works smoothly!

On the other hand, I find that `decoar2` uses the `TransformerEncoder` defined in **fairseq** and its forward function is expected to have one output variable (last layer representation), while the latest fairseq updated to have two output variables (last layer representation and `layer_result` for hidden states).
I guess the fairseq version you are using is the stable 0.10.2, while it contains some known bugs so we mostly use the latest master of fairseq in this codebase. Specifically I use this fairseq version: `8df9e3a4`.

Do you mind switch to the latest `TransformerEncoder` definition and make a few adjustment accordingly?
Thank you very much!

Sincerely,
Leo

File: upstream/decoar2/decoar2.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        x = self.dropout_input(features)
+
+        if padding_mask is not None:
+            extra = padding_mask.size(1) % features.size(1)
+            if extra > 0:
+                padding_mask = padding_mask[:, :-extra]
+            padding_mask = padding_mask.view(padding_mask.size(0), features.size(1), -1)
+            padding_mask = padding_mask.all(-1)
+
+        x = self.encoder(x, padding_mask=padding_mask)
```
Comment (leo19941227) [inline upstream/decoar2/decoar2.py:62]: The latest `TransformerEncoder` in fairseq will have two output variables in this line.

Comment (leo19941227) [reply upstream/decoar2/decoar2.py:62]: Ref: https://github.com/pytorch/fairseq/blob/b3491ae9d4c3eaa24292512cee7c21def713c535/fairseq/models/wav2vec/wav2vec2.py#L861-L867

**Comment by leo19941227** [comment]
Recently we have updated all the upstream to return hidden states for downstream to learn a weighted sum on them.
Do you think it will be nice to let the returning dictionary of UpstreamExpert's forward method to have the following two keys?

https://github.com/s3prl/s3prl/blob/39460b7ab8d19cd2fbe3406ed1073a5dbe96bfc0/upstream/example/expert.py#L50-L51

These two keys are not obvious to find in the current implementation since they are handled by **UpstreamBase** when I use hooks to get the hidden states.

https://github.com/s3prl/s3prl/blob/39460b7ab8d19cd2fbe3406ed1073a5dbe96bfc0/upstream/interfaces.py#L134-L135

But I am using hooks (and the code gets complicated) just to make things easier for some upstream and it is not required to use **UpstreamBase**. We can always directly include these two keys in UpstreamExpert's forward function.

Thank you very much!

**Comment by lingss0918** [comment]
> hidden

**Comment by lingss0918** [comment]
> Recently we have updated all the upstream to return hidden states for downstream to learn a weighted sum on them.
> Do you think it will be nice to let the returning dictionary of UpstreamExpert's forward method to have the following two keys?
> 
> https://github.com/s3prl/s3prl/blob/39460b7ab8d19cd2fbe3406ed1073a5dbe96bfc0/upstream/example/expert.py#L50-L51
> 
> These two keys are not obvious to find in the current implementation since they are handled by **UpstreamBase** when I use hooks to get the hidden states.
> 
> https://github.com/s3prl/s3prl/blob/39460b7ab8d19cd2fbe3406ed1073a5dbe96bfc0/upstream/interfaces.py#L134-L135
> 
> But I am using hooks (and the code gets complicated) just to make things easier for some upstream and it is not required to use **UpstreamBase**. We can always directly include these two keys in UpstreamExpert's forward function.
> 
> Thank you very much!

This make sense. It's not a minor change to return all hidden states from model though and I'll try to update that part later. Thanks!

**Comment by leo19941227** [comment]
Hi @lingss0918!

Are you planning to add the all hidden states support in this PR or open another PR for that?
I think the basic function for DeCoAR is ready to merged if you like.
Thanks again for the great work!

**Comment by lingss0918** [comment]
> Hi @lingss0918!
> 
> Are you planning to add the all hidden states support in this PR or open another PR for that?
> I think the basic function for DeCoAR is ready to merged if you like.
> Thanks again for the great work!

Maybe let's merge first. I'll open another PR when it's ready. Thanks!

**Comment by leo19941227** [comment]
Sure! Many thanks!

**Pull request merged**